File: unity-editor-mcp/Core/Handshake.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace UnityEditorMCP.Core
{
    /// <summary>
    /// The connect-time handshake payload: protocol version + editor identity +
    /// the per-editor command availability list. This is what lets the server
    /// detect a protocol or wrong-project connection and advertise only the tools
    /// the connected editor actually supports (requirements B3/C3/C4/G7). A
    /// Unity-independent data carrier — the Unity layer fills UnityVersion,
    /// ProjectPath, and AvailableCommands.
    /// </summary>
    public sealed class Handshake
    {
        public string ProtocolVersion { get; set; } = CommandCatalog.ProtocolVersion;
        public string UnityVersion { get; set; }
        public string ProjectPath { get; set; }
        public IReadOnlyList<string> AvailableCommands { get; set; } = Array.Empty<string>();

        public string ToJson()
        {
            var commands = new JArray();
            foreach (var c in AvailableCommands ?? Array.Empty<string>()) commands.Add(c);
            var o = new JObject
            {
                ["protocolVersion"] = ProtocolVersion,
                ["unityVersion"] = UnityVersion,
                ["projectPath"] = ProjectPath,
                ["availableCommands"] = commands,
            };
            return o.ToString(Formatting.None);
        }

        public static Handshake FromJson(string json)
        {
            if (string.IsNullOrWhiteSpace(json)) throw new ArgumentException("empty handshake json", nameof(json));
            var o = JObject.Parse(json);
            return new Handshake
            {
                ProtocolVersion = (string)o["protocolVersion"],
                UnityVersion = (string)o["unityVersion"],
                ProjectPath = (string)o["projectPath"],
                AvailableCommands = o["availableCommands"] is JArray arr
                    ? arr.Select(t => (string)t).Where(s => s != null).ToList()
                    : new List<string>(),
            };
        }

        /// <summary>
        /// Checks this (remote) handshake against the local protocol version and,
        /// optionally, an expected project path. Returns a structured result the
        /// server uses to refuse a mismatched protocol or a wrong-project editor.
        /// </summary>
        public CompatibilityResult CheckAgainst(string localProtocolVersion, string expectedProjectPath = null)
        {
            var version = ProtocolCompatibility.Check(localProtocolVersion, ProtocolVersion);
            if (!version.Compatible) return version;

            if (!string.IsNullOrEmpty(expectedProjectPath) && !PathsEqual(expectedProjectPath, ProjectPath))
            {
                return new CompatibilityResult(false, "PROJECT_PATH_MISMATCH",
                    $"Connected editor project '{ProjectPath}' does not match the targeted '{expectedProjectPath}'.");
            }

            return version;
        }

        private static bool PathsEqual(string a, string b)
        {
            if (a == null || b == null) return false;
            return string.Equals(Normalize(a), Normalize(b), StringComparison.OrdinalIgnoreCase);
        }

        private static string Normalize(string path) => path.Replace('\\', '/').TrimEnd('/');
    }
}

File: mcp-server/src/core/handshake.js
/**
 * Connect-time handshake evaluation — the Node mirror of the C# Core
 * ProtocolCompatibility + Handshake.CheckAgainst (unity-editor-mcp/Core/). The
 * editor answers a `handshake` command with { protocolVersion, unityVersion,
 * projectPath, availableCommands }; the server checks it here and warns (or, in
 * future, refuses) on a protocol-major or wrong-project mismatch.
 *
 * Compatibility rule: peers are compatible iff they share a protocol major; a
 * minor skew within a major is tolerated.
 */

// Must match protocol/VERSION (verified by protocol/scripts/check-drift.mjs).
export const PROTOCOL_VERSION = '1.0.0';

/** Parses "MAJOR.MINOR.PATCH"; returns null if malformed. */
export function parseSemver(version) {
  if (typeof version !== 'string') return null;
  const match = version.trim().match(/^(\d+)\.(\d+)\.(\d+)$/);
  if (!match) return null;
  return { major: Number(match[1]), minor: Number(match[2]), patch: Number(match[3]) };
}

/** Compares two protocol versions; mirrors C# ProtocolCompatibility.Check. */
export function checkProtocolCompatibility(localVersion, remoteVersion) {
  const local = parseSemver(localVersion);
  const remote = parseSemver(remoteVersion);

  if (!local || !remote) {
    return {
      compatible: false,
      code: 'PROTOCOL_VERSION_MISMATCH',
      message: `Unparseable protocol version (local '${localVersion}', remote '${remoteVersion}').`,
    };
  }

  if (local.major !== remote.major) {
    return {
      compatible: false,
      code: 'PROTOCOL_VERSION_MISMATCH',
      message: `Protocol major mismatch (local ${localVersion}, remote ${remoteVersion}). ` +
        'Update whichever package is older so both share a protocol major.',
    };
  }

  return {
    compatible: true,
    code: null,
    message: local.minor === remote.minor
      ? `Protocol ${localVersion} matches.`
      : `Protocol minor skew (local ${localVersion}, remote ${remoteVersion}); compatible using the lower minor's surface.`,
  };
}

/**
 * Normalizes a project path for the PROJECT_PATH_MISMATCH check. Mirrors C#
 * Handshake.Normalize (separator + trailing-slash) plus the OrdinalIgnoreCase
 * comparison it uses — folded here as ASCII-only (same as discovery.js
 * normalizeProjectPath), so non-ASCII paths don't diverge the way JS
 * .toLowerCase() (full-Unicode, locale-sensitive) would.
 */
function normalizePath(path) {
  if (!path) return '';
  return path
    .replace(/\\/g, '/')
    .replace(/\/+$/, '')
    .replace(/[A-Z]/g, (c) => String.fromCharCode(c.charCodeAt(0) + 32));
}

/**
 * Evaluates a handshake payload against the local protocol version and, when
 * provided, an expected project path. Mirrors C# Handshake.CheckAgainst.
 * @returns {{compatible: boolean, code: string|null, message: string}}
 */
export function evaluateHandshake(handshake, options = {}) {
  const localProtocolVersion = options.localProtocolVersion || PROTOCOL_VERSION;
  const expectedProjectPath = options.expectedProjectPath || null;

  if (!handshake || typeof handshake !== 'object') {
    return { compatible: false, code: 'PROTOCOL_VERSION_MISMATCH', message: 'No handshake payload received.' };
  }

  const versionResult = checkProtocolCompatibility(localProtocolVersion, handshake.protocolVersion);
  if (!versionResult.compatible) return versionResult;

  if (expectedProjectPath && normalizePath(expectedProjectPath) !== normalizePath(handshake.projectPath)) {
    return {
      compatible: false,
      code: 'PROJECT_PATH_MISMATCH',
      message: `Connected editor project '${handshake.projectPath}' does not match the targeted '${expectedProjectPath}'.`,
    };
  }

  return versionResult;
}

/**
 * Performs the connect-time handshake against a connected editor: sends the
 * `handshake` command, evaluates the reply, and returns a resilient result that
 * NEVER throws — an editor predating the command (UNKNOWN_COMMAND) or any I/O
 * failure degrades to a non-fatal `performed:false`, so the connection is never
 * broken by the handshake itself.
 * @returns {Promise<{performed:boolean, reason?:string, compatible?:boolean,
 *   code?:string|null, message:string, handshake?:object}>}
 */
export async function performHandshake(connection, options = {}) {
  const expectedProjectPath = options.expectedProjectPath
    ?? (typeof process !== 'undefined' ? process.env.UNITY_PROJECT_PATH : null)
    ?? null;
  const localProtocolVersion = options.localProtocolVersion || PROTOCOL_VERSION;

  let handshake;
  try {
    handshake = await connection.sendCommand('handshake');
  } catch (err) {
    if (err && err.code === 'UNKNOWN_COMMAND') {
      return { performed: false, reason: 'unsupported', message: 'Editor predates the handshake command.' };
    }
    return { performed: false, reason: 'error', message: err ? err.message : 'handshake failed' };
  }

  const verdict = evaluateHandshake(handshake, { expectedProjectPath, localProtocolVersion });
  return { performed: true, handshake, ...verdict };
}
